lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -27,9 +39,94 @@ static char rtt_buffer[CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT_BUF_SIZE];
 
 static bool is_transport_initialized;
 
+struct trace_data_t {
+	void *fifo_reserved; /* 1st word reserved for use by fifo */
+	const uint8_t * const data;
+	uint32_t len;
+};
+
+K_FIFO_DEFINE(trace_fifo);
+
+#define TRACE_THREAD_STACK_SIZE 256
+#define TRACE_THREAD_PRIORITY CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PRIO
+
+void trace_handler_thread(void)
+{
+	while (1) {
+		struct trace_data_t *trace_data = k_fifo_get(&trace_fifo, K_FOREVER);
+		const uint8_t * const data = trace_data->data;
+		const uint32_t len = trace_data->len;
+
 #ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART
+		/* Split RAM buffer into smaller chunks to be transferred using DMA. */
+		const uint32_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
+		uint32_t remaining_bytes = len;
+		nrfx_err_t err;
+
+		while (remaining_bytes) {
+			size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
+			uint32_t idx = len - remaining_bytes;
+
+			if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
+				LOG_WRN("UARTE TX not available!");
+				break;
+			}
+			err = nrfx_uarte_tx(&uarte_inst, &data[idx], transfer_len);
+			if (err != NRFX_SUCCESS) {
+				LOG_ERR("nrfx_uarte_tx error: %d", err);
+				k_sem_give(&tx_sem);
+				break;
+			}
+			remaining_bytes -= transfer_len;
+		}
+		/* Wait for last UART transfer to finish */
+		k_sem_take(&tx_sem, K_FOREVER);
+		k_sem_give(&tx_sem);
```

Comment:
The UART done interrupt is giving the semaphore also (see `uarte_callback`). We're taking and giving it here only because we want to wait for the TX to finish before calling the callback. Before we called `nrf_modem_trace_processed_callback` from the UART interrupt, but then it would be executed in interrupt context which caused issues.

In addition the callback expects to get the same buffer pointers and lengths as was sent from libmodem before, and this ensures that we call the callback with the original pointer and length instead of smaller chunks in case the UART peripheral doesn't support transfering the full buffer with EasyDMA.

Reply:
Thank you explaining. Perhaps it would be a bit clearer if you took the semaphore before the entering the `while` loop and again at its end? You would avoid having to take and give the sem after transmitting all the bytes.
```
k_sem_take(tx, K_FOREVER);
while (remaining_bytes) {
    uart_tx();
    k_sem_take(tx, K_FOREVER);
}
```

Reply:
Yes that would be clearer.

Reply:
@lemrey Aren't we just moving the problem then? So instead of having an out of place take & give at the end, we have an out of place take at the beginning and change to an uncommon order of take and the resource it's protecting (when reading the while-loop contents). I think coming back and trying to understand that code in a few months will be just as confusing.

How about just extracting the take and give to a function with a descriptive name? `wait_for_uart_done`?

Reply:
The difference would be that thread is only taking the semaphore and the interrupt is only giving the semaphore.
The thread always waits for a resource to be given by the interrupt, and does not give that resource itself; for me this is clearer to understand. 

The thread takes the semaphore twice
- before making the first write, to make sure there are no ongoing writes
- after each write, to make sure the operation has completed before making the next

That's clearer than:
The thread takes the semaphore twice, 
- after each write, to make sure the operation has completed before making the next
- once after the last write, to make sure the call does not return before the operation has completed (conceptually the same purpose as above, but taken in a different place)

and gives it once:
- once after completing the last request, for which a semaphore was given, but we have taken it

Reply:
I tried your suggestion to move the `k_sem_take`'s and remove the `k_sem_give`, but got full modem library heap very quickly. This is because that the `k_sem_give` was removed for the last TX. It's easier to see if we imagine the loop only run once and remove the while:

```
// tx_sem initial value == 1
k_sem_take(tx, K_FOREVER);
// tx_sem == 0
uart_tx();
// tx_sem is taken, so waiting for tx to finish
k_sem_take(tx, K_FOREVER);
// tx_sem == 0
```

For the next trace, the `tx_sem` value will be 0, and it will wait forever for the semaphore to be given. Calling `k_sem_give` after the while loop solves the problem. This means we're back to a similar solution to the original code. 

I have updated the PR to use my previous suggestion to move the take and give to a function with a meaningful name. Please have a look. Do you have other ideas how to solve the problem?

Reply:
Thanks for trying it out and explaining; it's fine as it is then.